### ISSUE TITLE
make it possible to build only igvtools as a fat jar

### ISSUE THREAD

**Issue by scalavision** [igvteam/igv#763]
`igvtools` is really nice to have in a pipeline to create tdf tracks for wgs bam files. It feels a bit unnecessary to add all of `igv` just for this. Could it be possible to make a build command that could make this into a fat, standalone jar?

**Comment by jrobinso** [comment]
Possible I suppose, but not straightforward.  You would need some kind of tool that works with Java 11 that could pack all the code from the various libraries into a single jar, and presumably you would want to tree shake it to remove unused classes.  We had such a thing for Java 8,  but it no longer works.

**Comment by jrobinso** [comment]
@davideby what are the options, if any, for combining everything into a single jar?

@scalavision Even if its all in a jar, it won't necessarily be any smaller.  That would depend on what options if any there are for tree shaking with Java 11.

**Comment by davideby** [comment]
@jrobinso @scalavision Even back with Java 8, we already were unable to package everything into a single jar: remember the goby & batik jar exceptions?  So we were already at a minimum of three jars since at least 2.4.x and maybe earlier due to some peculiarities of how some of our dependencies are packaged.

Java 9 forced a big push for modularity - where modules are defined at the jar level - making it no-longer-recommended to mix and match by repackaging across modules.  Some of that kind of repackaging is definitely considered invalid while other kinds might technically work but skirt the boundaries.

That's not to say it's impossible, just more difficult.  We experimented with ProGuard as a tree-shaker during the Java 9 port but these things always require tuning (not necessarily simple).  I'm sure we could reduce the jar sizes for sure, but it's not easy.  As with everything, it's just a matter of effort and we decided against it at the time.

In any case, it's still likely to be at least three jars and probably more.  I don't know, maybe we could shake loose even more (batik in particular) in an igvtools context.  But it would take some analysis to figure it out.
